Add a guide on threat modeling (#42980)
mdn/content: files/en-us/web/security/threat_modeling/index.md

- **T**ransfer: Shift responsibility for mitigating the threat to another system or organization (for example, a third-party service).
- **A**ccept: that it is not possible to mitigate the threat for now; it is still open and needs to be monitored.

Examples:

- Eliminate: We remove the comment functionality from our blog as it is not widely used and we don't want to deal with securing it.
- Reduce: We only allow registered users to use the comment function.
- Transfer: We use an external plugin for comments.
- Accept: We accept that our comment functionality faces threats, like spam, but we accept this threat and we implement spam monitoring.

Document your responses and decisions. You will likely come back to them in step 4 when asking if these responses are good enough.

To reference the identified responses, you index them with the letter R (R1, R2, R3, ...) in your threat model.

## 4. Did we do a good enough job?

Once you have done a round of threat modeling, file (private) issues with your project and describe your findings in a threat model document. Even if no action or bugfix is needed, the documentation of your threat model will be useful later.

> You can revisit the issues you filed and the documentation you've written in the next round of threat modeling and see if anything changed or needs reassessment. It is helpful to revalidate your documented issues. With every iteration of threat modeling, your system should become more secure and you will be more aware of further threats and risks. The experience you gather over time will help you to make your threat modeling more robust; it doesn't need to be perfect or complete from the start.

Reviewer: ```suggestion
You can revisit the issues you filed and the documentation you've written in the next round of threat modeling and see if anything changed or needs reassessment. It is helpful to revalidate your documented issues. With every iteration of threat modeling, your system should become more secure and you will be more aware of further threats and risks. The experience you gather over time will help you to make your threat modeling more robust; it won't be perfect or complete from the start, and it doesn't need to be in order to be useful.
```

---
title: Threat modeling
slug: Web/Security/Threat_modeling
page-type: guide
sidebar: security
---

Threat modeling is a process that can help developers identify and understand potential security risks in applications and websites. Threat modeling can help you understand the specific vulnerabilities of your application, the browser environment, and the user's interaction with your UI. This article describes what a threat model is and how to perform threat modeling, providing a lightweight overview and walking through the threat modeling process.

> Depending on your goal, threat modeling can be more involved than described here. Whether you are exercising lightweight threat modeling for your own benefit, so that security isn't an afterthought or neglected at all, or performing a more heavyweight assessment for a software audit, a threat model enables identifying and resolving actual and perceived threats.

Reviewer: FWIW a very long sentence. Generally consider shorter sentences. Easier to read and review. FWIW here means "I can live with it, but I would cut it back".

For threat model frameworks and resources, see:

- [Threat model frameworks an resources](/en-US/docs/Web/Security/Threat_modeling/Frameworks)
  - : Overview of the STRIDE and LUNDDUN frameworks that provide structure for threat modeling processes, and additional threat modeling resources.

For an example threat model, see:

- [Example threat model](/en-US/docs/Web/Security/Threat_modeling/Example_threat_model)
  - : An example threat model for a public-facing blog serving static pages. Interactive components include user comments, a contact form, analytics scripts, and a map embed.

## What is a threat?

A threat is anything that could potentially harm your website's functionality, or the data it holds.

A threat model is a structured representation of potential threats. It includes all the information that affects the security of your product, whether that product is a server, an application or a website. It is a living document, or mental map, that identifies your assets (What are you protecting?), potential adversaries (who would want to attack you, your product, or your users?), and potential vulnerabilities (where and what are your product's weak spots?)

> Threats are always present but they don't have to turn into attacks. An attack is a realization of a threat on an operational system (a system is a collection of assets) and ideally, if the security posture of a system is strong enough, potential threats won't turn into actual attacks.

Reviewer: This all sounds very techy - does it need to?

```suggestion
Threats are always present but they don't have to turn into attacks. An attack is when a threat is actually carried out against a live system (a system being a collection of assets). Ideally, if a system is well-protected, threats will stay as threats and never actually happen.
```

1. What are we working on?
2. What can go wrong?
3. What are we going to do about it?
4. Did we do a good enough job?

## How to do threat modeling?

Threat modeling should happen early in your development process and frequently revisited. Much like you constantly iterate over your software, you should also continuously analyze the system's security using your threat model. Usually, this begins right after the features have been defined.

Modeling activities are not done by security auditors exclusively. Anyone who is concerned about the privacy or security of a system should be empowered to help. Cross-functional collaboration from a diverse set of participants makes the threat model stronger. For example, whoever is designing the system surely has a clear understanding of what is being built and of the concerns that might keep them up at night.

Having a shared/common understanding of your system and its threats allows you to measure the robustness of your system. This should be captured in a threat model document.

> It can be quite some work to get to an initial threat modeling document. Often, the initial work gets done as part of a workshop with your team, either self-guided or mediated by a professional. The produced threat model document needs to be extensible for future re-assessment and ideally lives version-controlled within your codebase.

Reviewer: ```suggestion
It can be quite some work to get to an initial threat modeling document. Often, this work gets done as part of a workshop with your team, either self-guided or mediated by a professional. The produced threat model document needs to be extensible for future re-assessment and ideally lives version-controlled within your codebase.
```

- Identify stakeholders
- Discuss threats
- Consider threat responses
- Iterate

There is no single ideal threat modeling representation, therefore it is a good idea to use multiple threat modeling frameworks (see below for more on some popular ones) to illuminate different problems. We will focus on a lightweight threat modeling exercise which focuses on the four main questions from the Threat Modeling Manifesto.

The rest of this page walks through these steps in order.

## 1. What are we working on?

To answer this question, you typically create a model of the system, such as a data flow diagram, architecture diagram, or use-case diagram, showing components, data flows, trust boundaries, dependencies, and key stakeholders.

To define the scope of the threat model, we need to delineate which threats target our own project and which target the browser or other layers of the web platform stack, which we can define are external dependencies to our threat model. The [Threat Model for the Web Platform](https://w3c.github.io/threat-model-web/) provides a useful starting point and gives an idea of the environment any website or web application resides in.

> It is helpful to make yourself aware about which parts you will be responsible for vs. which parts will be taken care of by others, such as protections the browser usually provides for you. If you maintain a list of relevant existing threat models for your software dependencies and your environment, you can reference them in your own threat model and don't need to redo the modeling again. Threat modeling is not about completeness; it's about improving understanding over time.

Reviewer: Apparently screen readers do not like latin abbeviations, or so I was told when I started here.

```suggestion
It is helpful to make yourself aware about which parts you will be responsible for and which parts will be taken care of by others, such as protections the browser usually provides for you. If you maintain a list of relevant existing threat models for your software dependencies and your environment, you can reference them in your own threat model and don't need to redo the modeling again. Threat modeling is not about completeness; it's about improving understanding over time.
```

---
title: Threat modeling
slug: Web/Security/Threat_modeling
page-type: guide
sidebar: security
---

This article describes how to perform threat modeling for a website or a web application. We will provide a lightweight overview of the threat modeling process.

Depending on your goal, threat modeling can be more involved than described here. Maybe you or your team wants to exercise lightweight threat modeling for your own benefit, so that security isn't an afterthought or neglected at all, or maybe you are required to perform a more heavyweight assessment because your software will be audited and you use threat modeling to be prepared for that.

> In any case, it always makes sense to think about threats and a process that regularly assesses the security of your systems. The goal of this article is to provide some inspiration for how to do threat modeling, but ultimately you'll have to introduce a process that works for you, your team, and your goals.

Author: I just tried to have this at the bottom but it doesn't flow as nicely, I think. I could imagine to have a summary section at the end of the article but not sure.

Documentation of the common understanding of your system in a threat model enables you to measure the robustness of your system. It can be quite some work to get to an initial threat modeling document. Often, the initial work gets done as part of a workshop with your team, either self-guided or mediated by a professional. The produced threat model document needs to be extensible for future re-assessment and ideally lives version-controlled within your codebase.

For any threat model, it helps to:

- Describe your system elements (assets, components)
- Describe data flows and interactions with third parties
- Identify stakeholders
- Discuss threats
- Consider threat responses
- Iterate

There is no single ideal threat modeling representation, therefore it is a good idea to use multiple threat modeling frameworks (see below for more on some popular ones) to illuminate different problems. We will focus on a lightweight threat modeling exercise which focuses on the four main questions from the Threat Modeling Manifesto.

The rest of this page walks through these steps in order.

## 1. What are we working on?

> To answer this question, you typically create a diagram, map out your system, or illustrate the individual components and assets your system consists of.

Reviewer 2: ```suggestion
To answer this question, you typically create a diagram, map out your system, or illustrate the individual components and assets your system consists of. For example, a Data Flow Diagram or swimlanes, but also UML.
```
